modules/distributor/forwarder.go

```diff
@@ -0,0 +1,311 @@
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_dropped_pushes",
+		Help:      "Total number of failed pushes to the queue for a tenant to the forwarder",
+	}, []string{"tenant"})
+	metricForwarderQueueLength = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"tenant"})
+)
+
+type forwardFunc func(ctx context.Context, tenantID string, keys []uint32, traces []*rebatchedTrace) error
+
+type request struct {
+	keys   []uint32
+	traces []*rebatchedTrace
+}
+
+// forwarder queues up traces to be sent to the metrics-generators
+type forwarder struct {
+	services.Service
+
+	// per-tenant queue managers
+	queueManagers map[string]*queueManager
+	mutex         sync.Mutex
+
+	forwardFunc forwardFunc
+
+	o                 *overrides.Overrides
+	overridesInterval time.Duration
+	shutdown          chan interface{}
+}
+
+func newForwarder(fn forwardFunc, o *overrides.Overrides) *forwarder {
+	rf := &forwarder{
+		queueManagers:     make(map[string]*queueManager),
+		mutex:             sync.Mutex{},
+		forwardFunc:       fn,
+		o:                 o,
+		overridesInterval: time.Minute, // TODO: Make this configurable
+		shutdown:          make(chan interface{}),
+	}
+
+	rf.Service = services.NewIdleService(rf.start, rf.stop)
+
+	return rf
+}
+
+// SendTraces queues up traces to be sent to the metrics-generators
+func (f *forwarder) SendTraces(ctx context.Context, tenantID string, keys []uint32, traces []*rebatchedTrace) {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	select {
+	case <-f.shutdown:
+		return
+	default:
+	}
+
+	if _, ok := f.queueManagers[tenantID]; !ok {
+		queueSize := f.o.MetricsGeneratorForwarderQueueSize(tenantID)
+		workerCount := f.o.MetricsGeneratorForwarderWorkers(tenantID)
+		f.queueManagers[tenantID] = newQueueManager(tenantID, queueSize, workerCount, f.forwardFunc)
+	}
+
+	err := f.queueManagers[tenantID].pushToQueue(ctx, &request{keys: keys, traces: traces})
+	metricForwarderPushes.WithLabelValues(tenantID).Inc()
+	if err != nil {
+		// TODO: we may want to log the err
+		metricForwarderDroppedPushes.WithLabelValues(tenantID).Inc()
+	}
+}
+
+// watchOverrides watches the overrides for changes
+// and updates the queueManagers accordingly
+func (f *forwarder) watchOverrides() {
+	for {
+		select {
+		case <-time.After(f.overridesInterval):
+			f.mutex.Lock()
+			for tenantID, tm := range f.queueManagers {
+				queueSize := f.o.MetricsGeneratorForwarderQueueSize(tenantID)
+				workerCount := f.o.MetricsGeneratorForwarderWorkers(tenantID)
+
+				// if the queue size or worker count has changed, shutdown the queue manager and create a new one
+				if tm.shouldUpdate(queueSize, workerCount) {
+					go func() {
+						// shutdown the queue manager
+						// this will block until all workers have finished and the queue is drained
+						if err := tm.shutdown(); err != nil {
```

Comment:
Good points.

> Shutdown will mark the queue as read-only and discard any new pushes (= guaranteed data loss).

Data loss is protected by the mutex. Notice that the lock is taken during this loop and while pushing requests to the queue. It can't happen that a push is tried and a `queueManager` isn't ready to accept writes.

And even then, reloads here should be very fast. If a `queueManager` needs update, we're deleting from the map and creating a new one. Waiting for the old one to get drained happens in an independent goroutine.

> Should we instead have a dedicated function that does reloading?

Ideally, I agree. I tried reloading `queueManager`s instead of deleting&recreating, but the logic became way more complicated and less stable.

Having to drain a channel to close it and create a new one, while still accepting writes and having workers consume from _dynamic_ channels - I found it very tricky to manage. Personally, I'd keep the current approach unless it becomes a problem and doesn't solve our needs properly.